bpf: Refactor BPF datapath to cleanly separate static data from preprocessor conditionals

joestringer opened the pull request:
Refactor use of preprocessor macros that define static data such as IPs, LXC ids, etc. such that they are only used within datapath code as data values, and not to conditionally enable code paths.

Related: #6618

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6681)
<!-- Reviewable:end -->

joestringer committed:
bpf: Use ENABLE_IPV* for preprocessor branches

Allow the LXC_IP, LXC_IPV4 macros to be used purely as a piece of static
data rather than as a basis for enabling code branches.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
bpf: Compile-test with host redirect enabled

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
bpf: Split local delivery metrics macro from LXC_ID

In an effort to convert all usage of LXC_ID into static configuration
data usage, replace the current usage in metrics with a new macro that
will always be defined for all endpoint programs, LOCAL_DELIVERY_METRICS.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
bpf: Remove LXC_ID_NB

The network byte-order version of LXC_ID was never used, because an
LXC_ID is only meaningful within the local node. Remove its definition.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
bpf: Remove POLICY_ID

This macro was never used.

Signed-off-by: Joe Stringer <joe@cilium.io>

ianvernon commented on pkg/endpoint/bpf.go:
```diff
@@ -219,6 +219,9 @@ func (e *Endpoint) writeHeaderfile(prefix string, owner Owner) error {
 	fw.WriteString("#define LB_L4\n")
 
+	// Local delivery metrics should always be set for endpoint programs.
+	fw.WriteString("#define LOCAL_DELIVERY_METRICS\n")
```
Does this new flag have to be added to the `lxc_config.h` for the verifier testing?

ianvernon reviewed (commented):
LGTM from janitor / endpoint side.

tgraf commented on bpf/lib/eps.h:
```diff
@@ -98,7 +98,6 @@ ipcache_lookup4(struct bpf_elf_map *map, __be32 addr, __u32 prefix)
 	return map_lookup_elem(map, &key);
 }
 
-#if defined LXC_ID || defined SOCKMAP
```
Please explain in the commit message why removal of the `#if` is OK

joestringer replied:
From the new commit message:

```
    The wrappers around lookup_ip*_remote_endpoint() wrap the definition of
    macros to rewrite lookups into the ipcache map. In the non-LPM case,
    the IPCACHE_PREFIXES macro must be defined to ensure that the lookups
    find all ipcache entries; in the LPM case, so long as the macro is being
    used it must be defined. If it is defined and not used, the compiler
    should compile out the unnecessary code.
    
    The policy_can_egress*() functions only called from the egress BPF
    program which is attached to lxc devices, so the LXC_ID check there was
    a no-op.
    
    These changes should have no functional impact.
```

joestringer committed:
bpf: Remove unnecessary LXC_ID references

The wrappers around lookup_ip*_remote_endpoint() wrap the definition of
macros to rewrite lookups into the ipcache map. In the non-LPM case,
the IPCACHE_PREFIXES macro must be defined to ensure that the lookups
find all ipcache entries; in the LPM case, so long as the macro is being
used it must be defined. If it is defined and not used, the compiler
should compile out the unnecessary code.

The policy_can_egress*() functions only called from the egress BPF
program which is attached to lxc devices, so the LXC_ID check there was
a no-op.

These changes should have no functional impact.

Signed-off-by: Joe Stringer <joe@cilium.io>
